cardano-api/src/Cardano/Api/Fees.hs

```diff
@@ -263,6 +290,218 @@ estimateTransactionKeyWitnessCount TxBodyContent {
       _ -> 0
 
 
+-- ----------------------------------------------------------------------------
+-- Script execution units
+--
+
+-- | The different possible reasons that executing a script can fail,
+-- as reported by 'evaluateTransactionExecutionUnits'.
+--
+-- The first three of these are about failures before we even get to execute
+-- the script, and two are the result of execution.
+--
+data ScriptExecutionError =
+
+       -- | The script depends on a 'TxIn' that has not been provided in the
+       -- given 'UTxO' subset. The given 'UTxO' must cover all the inputs
+       -- the transaction references.
+       ScriptErrorMissingTxIn TxIn
+
+       -- | The 'TxIn' the script is spending does not have a 'ScriptDatum'.
+       -- All inputs guarded by Plutus scripts need to have been created with
+       -- a 'ScriptDatum'.
+     | ScriptErrorTxInWithoutDatum TxIn
+
+       -- | The 'ScriptDatum' provided does not match the one from the 'UTxO'.
+       -- This means the wrong 'ScriptDatum' value has been provided.
+       --
+     | ScriptErrorWrongDatum (Hash ScriptData)
```

Comment:
I think it would be worth including the incorrect and correct datum.

Reply:
I agree, but I don't think we have that immediately available. We'd probably need to change the ledger functions too.